src/ring_theory/integral_closure.lean

```diff
@@ -297,6 +297,37 @@ lemma algebra.is_integral_trans (A_int : ∀ x : A, is_integral R x)(B_int : ∀
   ∀ x:B, is_integral R x :=
 λ x, is_integral_trans A_int x (B_int x)
 
+lemma is_integral_of_surjective (h : function.surjective (algebra_map R A)) :
+  ∀ x : A, is_integral R x :=
+λ x, (h x).rec_on (λ y hy, (hy ▸ is_integral_algebra_map : is_integral R x))
+
+/-- If `R → A → B` is an algebra tower with `A → B` injective,
+then if the entire tower is an integral extension so is `R → A` -/
+lemma is_integral_of_is_integral_tower_bot (H : function.injective (algebra_map A B)) :
+  (∀ x : B, is_integral R x) → (∀ x : A, is_integral R x) :=
+begin
+  intros h x,
+  rcases h (algebra_map A B x) with ⟨p, ⟨hp, hp'⟩⟩,
+  refine ⟨p, ⟨hp, _⟩⟩,
+  rw [aeval_def, is_algebra_tower.algebra_map_eq R A B, ← eval₂_map,
+      eval₂_hom, ← ring_hom.map_zero (algebra_map A B)] at hp',
+  rw [aeval_def, eval₂_eq_eval_map],
+  exact H hp',
+end
+
+/-- If `R → A → B` is an algebra tower,
+then if the entire tower is an integral extension so is `A → B` -/
+lemma is_integral_of_is_integral_tower_top :
+  (∀ x : B, is_integral R x) → (∀ x : B, is_integral A x) :=
+begin
+  intros h x,
```

Comment:
I think this statement is true pointwise, for every `x : B`, right? so you can do without the two `forall` statements. We might want both versions.

Reply:
Yes point-wise versions work, and the point-wise version of `is_integral_tower_bot_of_is_integral` is (slightly) stronger than the version that I had originally, so I think having the point-wise versions makes more sense. Maybe if we had a definition of `integral_extension B R := (∀ x : B, is_integral R x)` then the `forall` versions might make sense to have, but they might still be unnecessary.